flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,240 @@
+        javaAgents = objects.domainObjectContainer(JavaAgent.class);
+        from(getProject().tarTree(LauncherResource.instance), copySpec -> exclude(JarFile.MANIFEST_NAME));
+
+        Provider<File> heartbeatJarProvider = getProject().provider(() -> {
+            File dest = new File(FlaskJarTask.this.getTemporaryDir(), HeartbeatAgentResource.instance.getDisplayName());
+            try (OutputStream os = Flask.write(dest, true); InputStream is = HeartbeatAgentResource.instance.read()) {
+                Flask.write2Stream(is, os);
+            }
+            return dest;
+        });
+        includeLibraries(heartbeatJarProvider);
+    }
+
+    @Input
+    public String getLauncherArchiveHash() {
+        return Flask.bytes2Hex(Flask.computeSHA256Digest(LauncherResource.instance::read));
+    }
+
+    @Input
+    public String getHeartbeatAgentHash() {
+        return Flask.bytes2Hex(Flask.computeSHA256Digest(HeartbeatAgentResource.instance::read));
+    }
+
+    @RequiredArgsConstructor
+    private static class StreamAction implements CopyActionProcessingStreamAction {
+
+        private final ZipOutputStream zoos;
+        private final Manifest manifest;
+        private final MessageDigest md;
+        private final byte[] buffer;
+
+        @Override
+        @SneakyThrows
+        public void processFile(FileCopyDetailsInternal fileCopyDetails) {
+            String entryName = fileCopyDetails.getRelativePath().toString();
+            if (!fileCopyDetails.isDirectory() && entryName.startsWith(Flask.Constants.LIBRARIES_FOLDER)) {
+                Supplier<InputStream> streamSupplier = () -> Flask.read(fileCopyDetails.getFile(), true);
+                Attributes attr = manifest.getEntries().computeIfAbsent(entryName, it -> new Attributes());
+                md.reset();
+                attr.putValue(Flask.ManifestAttributes.ENTRY_HASH,
+                        Base64.getEncoder().encodeToString(Flask.computeDigest(streamSupplier, md)));
+            }
+            if (Flask.Constants.METADATA_FOLDER.equals(entryName)) return;
+            ZipEntry zipEntry = new ZipEntry(entryName);
+            if (fileCopyDetails.isDirectory()) {
+                zipEntry = new ZipEntry(entryName + '/');
+                zoos.putNextEntry(zipEntry);
+            } else {
+                boolean compressed = Flask.splitExtension(fileCopyDetails.getSourceName())
+                        .map(entry -> ".jar".equals(entry.getValue()))
+                        .orElse(false);
+                if (!compressed) {
+                    zipEntry.setMethod(ZipEntry.DEFLATED);
+                } else {
+                    try (InputStream is = Flask.read(fileCopyDetails.getFile(), true)) {
+                        Flask.computeSizeAndCrc32(zipEntry, is, buffer);
+                    }
+                    zipEntry.setMethod(ZipEntry.STORED);
+                }
+                try (InputStream is = Flask.read(fileCopyDetails.getFile(), true)) {
```

Comment:
I suspect both these input streams can use the existing buffering provided by `computeSizeAndCrc32()` and `write2Stream()`.

My concern with the buffering was that I don't trust the Zip streams to perform File I/O directly.